### ISSUE TITLE
Can someone implement the equivalent of `torch.cat(...)` or `np.concatenate(...)` for grad tensors in TinyGrad?

### ISSUE THREAD

**Issue by guglielmocamporese** [geohot/tinygrad#274]
It would be easier for implementing Transformer based models (BERT, ViT, ...), for example for using the `CLS` and `SEP` tokens...

**Comment by sixChar** [comment]
Its probably far from efficient but would this work?

`def cat(tensors, dim=0):
    num_dims = len(tensors[0].shape)
    # So you can set dim=-1 for last dim
    if dim < 0:
        dim = dim + num_dims

    last_dim = num_dims-1

    # If dim is not last, transpose tensor so that the concat dim is last
    if dim != last_dim:
        order = np.arange(len(tensors[0].shape))
        order[dim] = last_dim
        order[last_dim] = dim
        tensors = [t.transpose(order=order) for t in tensors]
    
    sizes = [t.shape[-1] for t in tensors]
    # Size of result along concat dim
    total_size = sum(sizes)

    iden = Tensor.eye(total_size)
    # Start result as all zeros
    concat = Tensor.zeros(*tensors[0].shape[:-1], total_size)
    start = 0
    for i,t in enumerate(tensors):
        # dot each tensor with a slice of the identity so it has the same size as concat and it's values 
        # are in the right spots with zeros everywhere else
        concat = concat.add(t.dot(iden[start:start+sizes[i], :]))
        start += sizes[i]
    
    if dim != last_dim:
        # Undo transpose
        concat = concat.transpose(order=order)
    return concat`

**Comment by guglielmocamporese** [comment]
can you implement a test and compare your algo of `cat` wrt `torch` and `numpy` (in terms of execution time for example)?

**Comment by geohot** [comment]
Oh I like this!

**Comment by guglielmocamporese** [comment]
Hey, I tested this code for the cat, here some plots:

| Linear  | Log Scale |
| - | - | 
| <img width="497" alt="Screenshot 2021-06-29 at 1 27 08 AM" src="https://user-images.githubusercontent.com/31989563/123715772-4aa67f00-d879-11eb-9478-e075b8681a5a.png"> | <img width="497" alt="Screenshot 2021-06-29 at 1 27 15 AM" src="https://user-images.githubusercontent.com/31989563/123715776-4d08d900-d879-11eb-8474-4c7c47e4c212.png"> |

in this test I compared the cat function of `tinygrad` (above), `numpy` and `torch` by varying the inputs sizes. Here an idea of the code used:
```python
def measure_time(func, *args, **kwargs):
  to = time.time()
  out = func(*args, **kwargs)
  return out, time.time() - t0

# tinygrad
x1, x2 = Tensor.randn(1, size), Tensor.randn(1, size)
x, t = measure_time(cat, [x1, x2], dim=1) # your function above

# numpy
x1, x2 = np.random.randn(1, size), np.random.randn(1, size)
x, t = measure_time(np.concatenate, [x1, x2], dim=1)

# torch
x1, x2 = torch.randn(1, size), torch.randn(1, size)
x, t = measure_time(torch.cat, [x1, x2], dim=1)
```
In the plots the measurements are collected `num_trials`=10 times, and I highlighted the `min` and the `max` (colored region) and the `mean` (line).

It seems that your version has a 'squared' dependency respect to the input size, whereas it seems constant for `numpy` and `torch`...

How can we improve that?

**Comment by sixChar** [comment]
I think the backward pass would probably be a lot worse than pytorch's cat too.

I don't know if improving on my algorithm is the way to go. The main concern when writing it was to avoid using anything other than the existing 14 ops. If you're concerned about performance it'd probably be better to just write a new cat op at the same level as the other ops (I think there's already a pull request working on this).

Do you need better performance though? Roughly how many tensors are you going to be concatenating in a real scenario?

**Comment by sixChar** [comment]
Warning about the code I wrote! The backward pass won't work with tensors of rank > 2. I think it's because of how the dot backward pass is implemented.

**Comment by geohot** [comment]
I agree this is a real issue. I think you can write this as a second order op, Slice (if you go bigger, it pads) + Add.

**Comment by geohot** [comment]
Done
